src/preferences/common/preference-service.ts

```diff
@@ -0,0 +1,133 @@
+    // remove<T>(preferenceName: string): T | undefined;
+
+}
+
+export interface PreferenceServiceClient {
+    onDidChangePreference(event: PreferenceChangedEvent): void
+}
+
+export class JsonPreferenceService implements PreferenceService {
+
+    // A minimal part of the interface.
+    protected prefs: { [key: string]: any } | undefined; // Preferences cache
+    protected readonly prefsChangedListeners: Emitter<PreferenceChangedEvent> = new Emitter();
+    protected readonly resolveUri: Promise<URI>;
+
+    constructor(
+        @inject(FileSystem) protected readonly fileSystem: FileSystem,
+        @inject(FileSystemWatcher) protected readonly watcher: FileSystemWatcher,
+        readonly preferencePath: string) {
+
+        this.resolveUri = fileSystem.getWorkspaceRoot().then(root =>
+            new URI(root.uri).appendPath(preferencePath)
+        );
+
+
+
+        watcher.onFileChanges(event => {
+            this.arePreferencesAffected(event).then((arePrefsAffected) => {
+                if (arePrefsAffected) {
+                    this.reconcilePreferences();
+                }
+            })
+        });
+
+        this.reconcilePreferences();
+    }
+
+    /**
+     * Checks to see if the preference file was modified
+     */
+    arePreferencesAffected(event: FileChangesEvent): Promise<boolean> {
+        return this.resolveUri.then(uri =>
+            event.changes.some(c =>
+                c.uri === uri.toString()
+            )
+        );
+    }
+
+    /**
+     * Read preferences
+     */
+    reconcilePreferences(): void {
+
+        this.resolveUri.then((uri) => {
+            // const uri = path.resolve(root, "/", this.preferencePath);
+            this.fileSystem.resolveContent(uri.toString()).then(({ stat, content }) => {
+                let newPrefs = JSON.parse(content) // Might need a custom parser because comments and whatnot?
+                const newKeys: { [key: string]: any } = Object.keys(newPrefs).sort();
+                if (this.prefs !== undefined && this.prefs !== newPrefs) {
+                    const keys: { [key: string]: any } = Object.keys(this.prefs).sort();
+
+                    newKeys.forEach((newKey: string) => {
+                        if (keys.indexOf(newKey) !== -1) { // existing key
+                            if (newKeys[newKey] !== keys[newKey]) { // value changed
+                                const event: PreferenceChangedEvent = { preferenceName: newKey, newValue: newKeys[newKey], oldValue: keys[newKey] };
+                                this.prefsChangedListeners.fire(event);
+                            }
+                        }
+                    });
+                }
+                this.prefs = newPrefs;
+                newKeys.forEach((newKey: string) => {
+                    const event: PreferenceChangedEvent = { preferenceName: newKey, newValue: newKeys[newKey] };
+                    this.prefsChangedListeners.fire(event);
```

Comment:
There's probably a much better/cleaner way to get the differences and fire events for that. WIP and open to suggestions/comments.